qa/os/src/test/java/org/elasticsearch/packaging/test/DebPreservationTests.java

```diff
@@ -90,4 +93,31 @@ public void test30Purge() throws Exception {
 
         assertThat(packageStatus(distribution()).exitCode, is(1));
     }
+
+    /**
+     * Check that restarting on upgrade doesn't run into a problem where the keystore
+     * upgrade is attempted as the wrong user i.e. the restart happens at the correct
+     * point. See #82433.
+     */
+    public void test40RestartOnUpgrade() throws Exception {
+        assertRemoved(distribution());
+        installation = installPackage(sh, distribution());
+        assertInstalled(distribution());
+
+        // Ensure ES is started
+        sh.run("systemctl daemon-reload");
+        sh.run("systemctl enable elasticsearch.service");
+        sh.run("systemctl start elasticsearch.service");
+
+        final JournaldWrapper journaldWrapper = new JournaldWrapper(sh);
+        sh.getEnv().put("RESTART_ON_UPGRADE", "true");
+        installation = installPackage(sh, distribution());
+        final String logs = journaldWrapper.getLogs().stdout;
+
+        assertThat(
```

Comment:
This seems like an awfully specific assertion. Shouldn't we just assert that ES is successfully running when using that option. That would catch more regressions related to `RESTART_ON_UPGRADE` no?